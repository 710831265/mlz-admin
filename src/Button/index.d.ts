// / <reference types="react" />
import React from 'react';
import { SizeType } from 'antd/es/config-provider/SizeContext.d';
import { ProColumnType, ProTableTypes } from '@ant-design/pro-table/lib/Table.d';
import { OptionConfig } from '@ant-design/pro-table/lib/component/toolBar';
import { TablePaginationConfig } from 'antd/es/table/interface.d';
import { SearchConfig } from '@ant-design/pro-table/lib/Form';
import { ToolBarProps } from '@ant-design/pro-table/lib/component/toolBar/index.d';

declare namespace CommonButtonTypes {
  export type ButtonTypes = any;
}
