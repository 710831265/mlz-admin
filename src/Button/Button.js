import React from 'react';
import { Button as AntdButton, Dropdown, Menu } from 'antd';
import { default as AntdButtonGroup } from 'antd/es/button/button-group';
import { omitObject } from '@/shared/utils';
import Icon from '@/Icon/Icon';
import 'antd/es/button/style/index.css';
const calcValues = ($menuClickEvent, $menu) => {
  const identifier = 'no value/@mlz-admin/identifier';
  const values = $menu.reduce((prev, curr) => {
    return prev.concat([curr.key.toString() === $menuClickEvent.key ? curr.value : identifier]);
  }, []);
  const result = values.filter((item) => item !== identifier);
  return result.length === 0 ? undefined : result.length === 1 ? result[0] : result;
};
const Button = (props) => {
  const { group, menu } = props;
  return !group
    ? !menu
      ? React.createElement(AntdButton, Object.assign({}, props), props.children)
      : React.createElement(
          Dropdown,
          {
            overlay: React.createElement(
              Menu,
              { onClick: (e) => props.onChange && props.onChange(calcValues(e, menu)) },
              menu?.map((item, index) => {
                return React.createElement(Menu.Item, { key: item.key || index, icon: item.iconType ? React.createElement(Icon, { type: item?.iconType }) : undefined }, item.text);
              }),
            ),
          },
          React.createElement(Button, null, props.children),
        )
    : React.createElement(
        AntdButtonGroup,
        null,
        group?.map((item, index) => {
          return React.createElement(
            AntdButton,
            Object.assign({ key: item.key || index }, omitObject(props, ['size', 'style', 'className', 'prefixCls']), {
              onClick: (e) => {
                item.onClick && item.onClick(e);
                props.onChange && props.onChange(item.value || undefined);
              },
              type: item.type || props.type,
            }),
            item.leftIconType ? React.createElement(Icon, { type: item?.leftIconType }) : null,
            item.text,
            item.rightIconType ? React.createElement(Icon, { type: item?.rightIconType }) : null,
          );
        }),
      );
};
Button.Group = AntdButtonGroup;
export default Button;
//# sourceMappingURL=Button.js.map
