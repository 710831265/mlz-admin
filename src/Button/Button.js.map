{"version":3,"file":"Button.js","sourceRoot":"","sources":["Button.tsx"],"names":[],"mappings":"AAAA,OAAO,KAA+B,MAAM,OAAO,CAAC;AACpD,OAAO,EAAE,MAAM,IAAI,UAAU,EAAE,QAAQ,EAAE,IAAI,EAAE,MAAM,MAAM,CAAC;AAE5D,OAAO,EAAE,OAAO,IAAI,eAAe,EAAE,MAAM,6BAA6B,CAAC;AACzE,OAAO,EAAE,UAAU,EAAE,MAAM,gBAAgB,CAAC;AAC5C,OAAO,IAAI,MAAM,aAAa,CAAC;AAE/B,OAAO,gCAAgC,CAAC;AAExC,MAAM,UAAU,GAAG,CAAC,eAA2B,EAAE,KAAiB,EAAyD,EAAE;IAC3H,MAAM,UAAU,GAAG,gCAAgC,CAAC;IACpD,MAAM,MAAM,GAAG,KAAK,CAAC,MAAM,CAAC,CAAC,IAA2D,EAAE,IAAI,EAAE,EAAE;QAGhG,OAAO,IAAI,CAAC,MAAM,CAAC,CAAC,IAAI,CAAC,GAAG,CAAC,QAAQ,EAAE,KAAK,eAAe,CAAC,GAAG,CAAC,CAAC,CAAC,IAAI,CAAC,KAAK,CAAC,CAAC,CAAC,UAAU,CAAC,CAAC,CAAC;IAC9F,CAAC,EAAE,EAAE,CAAC,CAAC;IACP,MAAM,MAAM,GAAG,MAAM,CAAC,MAAM,CAAC,CAAC,IAA8B,EAAE,EAAE,CAAC,IAAI,KAAK,UAAU,CAAC,CAAC;IACtF,OAAO,MAAM,CAAC,MAAM,KAAK,CAAC,CAAC,CAAC,CAAC,SAAS,CAAC,CAAC,CAAC,MAAM,CAAC,MAAM,KAAK,CAAC,CAAC,CAAC,CAAC,MAAM,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,MAAM,CAAC;AACpF,CAAC,CAAC;AACF,MAAM,MAAM,GAAG,CAAC,KAAkB,EAAsB,EAAE;IACxD,MAAM,EAAE,KAAK,EAAE,IAAI,EAAE,GAAG,KAAK,CAAC;IAC9B,OAAO,CAAC,KAAK,CAAC,CAAC,CAAC,CACd,CAAC,IAAI,CAAC,CAAC,CAAC,CACN,oBAAC,UAAU,oBAAK,KAAK,GAAG,KAAK,CAAC,QAAQ,CAAc,CACrD,CAAC,CAAC,CAAC,CACF,oBAAC,QAAQ,IACP,OAAO,EACL,oBAAC,IAAI,IAAC,OAAO,EAAE,CAAC,CAAa,EAAE,EAAE,CAAC,KAAK,CAAC,QAAQ,IAAI,KAAK,CAAC,QAAQ,CAAC,UAAU,CAAC,CAAC,EAAE,IAAI,CAAC,CAAC,IACpF,IAAI,EAAE,GAAG,CAAC,CAAC,IAAI,EAAE,KAAK,EAAE,EAAE;YACzB,OAAO,CACL,oBAAC,IAAI,CAAC,IAAI,IAAC,GAAG,EAAE,IAAI,CAAC,GAAG,IAAI,KAAK,EAAE,IAAI,EAAE,IAAI,CAAC,QAAQ,CAAC,CAAC,CAAC,oBAAC,IAAI,IAAC,IAAI,EAAE,IAAI,EAAE,QAAQ,GAAI,CAAC,CAAC,CAAC,SAAS,IAChG,IAAI,CAAC,IAAI,CACA,CACb,CAAC;QACJ,CAAC,CAAC,CACG;QAET,oBAAC,MAAM,QAAE,KAAK,CAAC,QAAQ,CAAU,CACxB,CACZ,CACF,CAAC,CAAC,CAAC,CACF,oBAAC,eAAe,QACb,KAAK,EAAE,GAAG,CAAC,CAAC,IAAe,EAAE,KAAa,EAAE,EAAE;QAC7C,OAAO,CACL,oBAAC,UAAU,kBACT,GAAG,EAAE,IAAI,CAAC,GAAG,IAAI,KAAK,IAClB,UAAU,CAAC,KAAK,EAAE,CAAC,MAAM,EAAE,OAAO,EAAE,WAAW,EAAE,WAAW,CAAC,CAAC,IAClE,OAAO,EAAE,CAAC,CAAC,EAAE,EAAE;gBACb,IAAI,CAAC,OAAO,IAAI,IAAI,CAAC,OAAO,CAAC,CAAC,CAAC,CAAC;gBAChC,KAAK,CAAC,QAAQ,IAAI,KAAK,CAAC,QAAQ,CAAC,IAAI,CAAC,KAAK,IAAI,SAAS,CAAC,CAAC;YAC5D,CAAC,EACD,IAAI,EAAE,IAAI,CAAC,IAAI,IAAI,KAAK,CAAC,IAAI;YAC5B,IAAI,CAAC,YAAY,CAAC,CAAC,CAAC,oBAAC,IAAI,IAAC,IAAI,EAAE,IAAI,EAAE,YAAY,GAAI,CAAC,CAAC,CAAC,IAAI;YAC7D,IAAI,CAAC,IAAI;YACT,IAAI,CAAC,aAAa,CAAC,CAAC,CAAC,oBAAC,IAAI,IAAC,IAAI,EAAE,IAAI,EAAE,aAAa,GAAI,CAAC,CAAC,CAAC,IAAI,CACrD,CACd,CAAC;IACJ,CAAC,CAAC,CACc,CACnB,CAAC;AACJ,CAAC,CAAC;AAEF,MAAM,CAAC,KAAK,GAAG,eAAe,CAAC;AAC/B,eAAe,MAAM,CAAC"}