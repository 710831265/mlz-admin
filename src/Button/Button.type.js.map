{"version":3,"file":"Button.type.js","sourceRoot":"","sources":["Button.type.ts"],"names":[],"mappings":""}