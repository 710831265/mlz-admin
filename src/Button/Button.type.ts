// <reference types="react" />
import React from 'react';
import { ButtonProps, ButtonType } from 'antd/lib/button';

export type ButtonGroupPropableValue = any;
export interface GroupType {
  key: string | number;
  text: string;
  onClick?: React.MouseEventHandler<HTMLElement>;
  value?: ButtonGroupPropableValue;
  leftIconType?: string;
  rightIconType?: string;
  type?: ButtonType;
}
export type MenuType = { iconType?: string } & Exclude<GroupType, 'leftIconType' | 'rightIconType' | 'type' | 'onClick'>;

export interface ButtonTypes extends ButtonProps {
  menu?: MenuType[];
  group?: GroupType[];
  onChange?: (e: ButtonGroupPropableValue) => void;
}
export type ButtonGroupTypes = {
  children: React.ReactNode;
};
