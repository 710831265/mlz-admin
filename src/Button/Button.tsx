import React, { useState, useContext } from 'react';
import { Button as AntdButton, Dropdown, Menu } from 'antd';
import { ClickParam } from 'antd/lib/menu/index.d';
import { default as AntdButtonGroup } from 'antd/es/button/button-group';
import { omitProps } from 'mytils';
import Icon from '../Icon/Icon';
import { MenuType, GroupType, ButtonTypes, ButtonGroupPropableValue } from './Button.type';

const calcValues = ($menuClickEvent: ClickParam, $menu: MenuType[]): ButtonGroupPropableValue | ButtonGroupPropableValue[] => {
  const identifier = 'no value/@mlz-admin/identifier';
  const values = $menu.reduce((prev: ButtonGroupPropableValue | ButtonGroupPropableValue[], curr) => {
    // 必须使用特定的字符串，而不是null或undefined，因为
    // 用户可能想要的值就是它们
    return prev.concat([curr.key.toString() === $menuClickEvent.key ? curr.value : identifier]);
  }, []);
  const result = values.filter((item: ButtonGroupPropableValue) => item !== identifier);
  return result.length === 0 ? undefined : result.length === 1 ? result[0] : result;
};
const Button = (props: ButtonTypes): React.ReactElement => {
  const { group, menu } = props;
  return !group ? (
    !menu ? (
      <AntdButton {...props}>{props.children}</AntdButton>
    ) : (
      <Dropdown
        overlay={
          <Menu onClick={(e: ClickParam) => props.onChange && props.onChange(calcValues(e, menu))}>
            {menu?.map((item, index) => {
              return (
                <Menu.Item key={item.key || index} icon={item.iconType ? <Icon type={item?.iconType} /> : undefined}>
                  {item.text}
                </Menu.Item>
              );
            })}
          </Menu>
        }>
        <Button {...omitProps(['menu'], props)}>
          <Icon type="arrow_down" />
          {props.children}
        </Button>
      </Dropdown>
    )
  ) : (
    <AntdButtonGroup>
      {group?.map((item: GroupType, index: number) => {
        return (
          <AntdButton
            key={item.key || index}
            {...omitProps(['size', 'style', 'className', 'prefixCls'], props)}
            onClick={(e) => {
              item.onClick && item.onClick(e);
              props.onChange && props.onChange(item.value || undefined);
            }}
            type={item.type || props.type}>
            {item.leftIconType ? <Icon type={item?.leftIconType} /> : null}
            {item.text}
            {item.rightIconType ? <Icon type={item?.rightIconType} /> : null}
          </AntdButton>
        );
      })}
    </AntdButtonGroup>
  );
};

Button.Group = AntdButtonGroup;
export default Button;
