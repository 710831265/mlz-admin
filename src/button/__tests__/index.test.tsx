import React, { Component } from 'react';
import { mount } from 'enzyme';
import Button from '..';
import { testMount, testSnapshot } from '../../../tests';
const groupList = [
  {
    key: 1,
    text: '菜单1',
    value: 1,
    leftIconType: 'round_left_g',
  },
  {
    key: 2,
    text: '菜单2',
    value: 2,
    type: 'default',
  },
  {
    key: 3,
    text: '菜单3',
    value: 3,
    rightIconType: 'round_right_g',
  },
] as const;

describe('🧪 Button', () => {
  /**
   * @FIRST 挂载 / 快照测试（一定程度上防止圣诞节彩蛋等底层dom修改..）
   */
  testMount(Button);
  testSnapshot(Button);

  testMount(() => (
    <Button.Group>
      <Button>按钮1</Button>
      <Button>按钮2</Button>
      <Button>按钮3</Button>
    </Button.Group>
  ));

  /**
   * @SECOND 扩展属性
   */
  // type属性是否还可以控制按钮样式
  test('props.type rendered correctly', () => {
    ([
      {
        type: 'default',
        value: false,
      },
      {
        type: 'primary',
        value: true,
      },
    ] as const).forEach((opt) => {
      const wrapper = mount(<Button type={opt.type}>测试</Button>);
      expect(wrapper.find('.ant-btn').hasClass('ant-btn-primary')).toBe(opt.value);
    });
  });

  /**
   * @THIRD 事件/属性正确地触发/生效
   */
  test('onClick should be called within a composed params when using group prop', () => {
    const onClickFn = jest.fn();
    const wrapper = mount(<Button onClick={onClickFn} group={groupList} />);
    const btns = wrapper.find('.ant-btn');
    expect(btns.length).toBe(groupList.length);
    btns.at(0).simulate('click');
    expect(onClickFn).toHaveBeenCalled();
  });
});
