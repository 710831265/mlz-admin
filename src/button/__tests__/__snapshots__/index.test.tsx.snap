// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`🧪 Button snapshot test 快照匹配 1`] = `"<button type=\\"button\\" class=\\"ant-btn\\"></button>"`;
